Gracefully handle invalid synonym rules in updateable synonyms

Mikep86 opened the pull request:
Gracefully handle invalid synonym rules by setting `lenient` to `true` by default when synonyms are updateable. This allows us to handle scenarios like:

```json
PUT _synonyms/test-synonyms
{
  "synonyms_set": [
    {
      "synonyms": "foo, bar, baz"
    }
  ]
}

PUT test-index
{
  "settings":{
    "analysis": {
      "filter": {
        "stop_filter": {
          "type": "stop",
          "stopwords": ["bar"]
        },
        "synonym_filter": {
          "type": "synonym_graph",
          "synonyms_set": "test-synonyms",
          "updateable": true
        }
      },
      "analyzer": {
        "index_analyzer": {
          "tokenizer": "whitespace",
          "filter": [
            "stop_filter"
          ]
        },
        "search_analyzer": {
          "tokenizer": "whitespace",
          "filter": [
            "stop_filter",
            "synonym_filter"
          ]
        }
      }
    }
  },
  "mappings": {
    "properties": {
      "test_field": {
        "type": "text",
        "analyzer": "index_analyzer",
        "search_analyzer": "search_analyzer"
      }
    }
  }
}
```

Without any analyzer reload errors.

Mikep86 committed:
Added synonyms API integration test

Mikep86 committed:
Set cluster scope to TEST

Mikep86 committed:
Merge branch 'main' into handle-invalid-synonyms-sets

Mikep86 committed:
Default to "lenient: true" when using the Synonyms API

Mikep86 committed:
Remove unused Synonym token filter factories that were causing test failures

Mikep86 committed:
Test analysis with synonym analyzer

Mikep86 committed:
Test development

Mikep86 committed:
Don't create fake synonym rule

Mikep86 committed:
Test loading an index that references an empty synonyms set

Mikep86 commented:
@elasticmachine run elasticsearch-ci/bwc-snapshots

Mikep86 committed:
Merge branch 'main' into handle-invalid-synonyms-sets

Mikep86 commented on modules/analysis-common/src/internalClusterTest/java/org/elasticsearch/analysis/common/SynonymsApiIT.java:
```diff
@@ -0,0 +1,156 @@
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return List.of(CommonAnalysisPlugin.class, MapperExtrasPlugin.class, ReindexPlugin.class);
```
We need `MapperExtrasPlugin` and `ReindexPlugin` to support operations performed by [SynonymsManagementAPIService](https://github.com/elastic/elasticsearch/blob/d332ed7d16c69906ce8fd81d6ccbbef3300731a8/server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java)

Mikep86 commented on modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/SynonymTokenFilterFactory.java:
```diff
@@ -62,8 +76,10 @@ public class SynonymTokenFilterFactory extends AbstractTokenFilterFactory {
+        this.synonymsSource = parseSettingsForSynonymsSource(settings);
         this.expand = settings.getAsBoolean("expand", true);
-        this.lenient = settings.getAsBoolean("lenient", false);
+        this.lenient = settings.getAsBoolean("lenient", this.synonymsSource == SynonymsSource.INDEX);
```
This changes the default `lenient` value to `true` when using `synonyms_set`, but the user can still get into trouble if they explicitly set `lenient` to `false`. Do we want to throw an error on that?

carlosdelest replied:
I'd say to keep the setting as true by default in all types - that would be better from my PoV to avoid having users to understand that this setting depends on the type. And it's the safest thing to do for file based as well.

kderusso replied:
I like the idea of defaulting to `true` but I also think we could document this - that if `lenient` is set to `false` you may run into this behavior.

Mikep86 replied:
I'm going to set the default to true for all cases, barring any new feedback against this decision

Mikep86 commented on modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/SynonymTokenFilterFactory.java:
```diff
@@ -153,45 +169,80 @@ SynonymMap buildSynonyms(Analyzer analyzer, ReaderWithOrigin rules) {
+                // don't block a master thread
+                ReaderWithOrigin reader;
+                if (context != IndexCreationContext.RELOAD_ANALYZERS) {
+                    reader = new ReaderWithOrigin(
+                        new StringReader(""),
+                        "fake empty [" + synonymsSet + "] synonyms_set in .synonyms index",
+                        synonymsSet
+                    );
```
Changed the filter to initialize with an empty synonyms set instead of a fake rule. IMO this is better because now there is zero chance the user hits the fake rule in between init and when analyzers are reloaded. This empty synonyms set handling is tested in `SynonymsApiIT#testLoadIndexWithEmptySynonymsSet`.

Mikep86 commented on modules/analysis-common/src/test/java/org/elasticsearch/analysis/common/SynonymsAnalysisTests.java:
These `SynonymTokenFilterFactory` instances were causing test failures because the index settings don't configure a synonyms source, which is now checked at construction time. I fixed the tests by removing the offending token filters, which should be fine because they were unused.

Mikep86 commented on modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/SynonymTokenFilterFactory.java:
```diff
@@ -62,8 +76,10 @@ public class SynonymTokenFilterFactory extends AbstractTokenFilterFactory {
             );
         }
+
+        this.synonymsSource = parseSettingsForSynonymsSource(settings);
```
The synonyms source is now determined in the constructor so that all of the token filter's instance variables can remain `final`

elasticsearchmachine commented:
Pinging @elastic/es-search-relevance (Team:Search Relevance)

Mikep86 committed:
Update docs/changelog/110901.yaml

elasticsearchmachine commented:
Pinging @elastic/search-relevance (Team:Search - Relevance)

elasticsearchmachine commented:
Pinging @elastic/ent-search-eng (Team:SearchOrg)

elasticsearchmachine commented:
Hi @Mikep86, I've created a changelog YAML for you.

Mikep86 commented:
Can/should we backport this change to 8.15?

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

Mikep86 commented:
@carlosdelest 

> Is this a breaking change, as we're changing the default for a token filter setting?

I don't think so, as changing this default won't cause existing indices that use `synonyms_set` to break upon application.

> Should we apply the same defaults to, at least, the file based synonyms? Can't we have a similar problem there?

We could have a similar problem if the synonyms file changes to include an invalid synonym rule and then the index is reloaded. I think we could expand the default `lenient: true` to file-based synonyms if other reviewers agree.

carlosdelest commented:
> I don't think so, as changing this default won't cause existing indices that use synonyms_set to break upon application.

That is correct - my concern is about expectations from users, as we're changing defaults. I think it's better for them to find unintended consequences (a synonym is not loaded) vs potentially having a red index.

However, we're not signaling the users that they have problems in their synonym rules - we're just ignoring them. While I feel it's safer to do that than our current approach, I'd like to understand if we can provide some kind of error / warning (as a follow up).

> I think we could expand the default lenient: true to file-based synonyms if other reviewers agree.

I would change the default for all synonym types so users don't have to think what type they're using - we're taking the safest approach by default, and they're free to double check by using the `lenient: false` approach.

Mikep86 committed:
Fix changelog YAML

carlosdelest commented on modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/SynonymTokenFilterFactory.java:
```diff
@@ -33,6 +33,19 @@
 import java.util.function.Function;
 
 public class SynonymTokenFilterFactory extends AbstractTokenFilterFactory {
+    private enum SynonymsSource {
```
I've found [this enum pattern](https://github.com/elastic/elasticsearch/blob/9ba5651e7478216f17d81b63385ddf6db2331f7f/server/src/main/java/org/elasticsearch/Build.java#L152) more prominent in the codebase - using a string ctor and a displayName() method for retrieving it

Mikep86 replied:
Much better way to do it 🤦

carlosdelest commented on modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/SynonymTokenFilterFactory.java:
```diff
@@ -153,45 +169,80 @@ SynonymMap buildSynonyms(Analyzer analyzer, ReaderWithOrigin rules) {
-            StringBuilder sb = new StringBuilder();
-            for (String line : rulesList) {
-                sb.append(line).append(System.lineSeparator());
+        return switch (synonymsSource) {
```
I like the structure using enum. Should we take that a level further, and make the enums have a method to return the ReaderWithOrigin so we avoid the switch here?

carlosdelest commented on modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/SynonymTokenFilterFactory.java:
```diff
@@ -153,45 +169,80 @@ SynonymMap buildSynonyms(Analyzer analyzer, ReaderWithOrigin rules) {
         }
     }
+
+    private static SynonymsSource parseSettingsForSynonymsSource(Settings settings) {
```
This could be part of the enum class

carlosdelest commented on modules/analysis-common/src/internalClusterTest/java/org/elasticsearch/analysis/common/SynonymsApiIT.java:
Good set of tests - I like that we're being explicit on the cases we use in the docs 👏

Mikep86 commented:
> However, we're not signaling the users that they have problems in their synonym rules - we're just ignoring them. While I feel it's safer to do that than our current approach, I'd like to understand if we can provide some kind of error / warning (as a follow up).

Understood and agreed. We aren't completely silently ignoring the error, it is logged in the ES logs when `lenient: true`:

```
[2024-07-16T14:38:48,335][INFO ][o.e.a.c.ESSolrSynonymParser] [runTask-0] Synonym rule for [bar] was ignored
```

However, that is easy to miss. Better to surface this error through an API and make it visible through Kibana. I think this is part of a larger problem we have; asynchronous/background index errors aren't brought to the user's attention. Good thing to address more comprehensively in a follow-up task.

carlosdelest reviewed (commented):
Looks really good! 👏 

- Some comments on making the enum more first class and similar to what we have in other parts of the codebase.
- Would we benefit from having YAML tests? 🤔 
- I'd like others to chime in for the change in defaults, to ensure we're not putting ourselves into a corner

Mikep86 commented:
@carlosdelest 

> Would we benefit from having YAML tests?

I think `SynonymsApiIT` can be re-written as YAML tests, assuming we don't want to test file-based synonyms. AFAIK there's no way to create/delete/modify files using the YAML test framework.

kderusso commented on docs/changelog/110901.yaml:
```diff
@@ -0,0 +1,5 @@
+pr: 110901
+summary: Gracefully handle invalid synonym rules in synonyms sets
```
Nitpick: Do we need to mention anything about setting the default of lenient in the notes here?

Mikep86 replied:
Good catch, we can update the summary to mention changes to default values once it's decided what those changes are.

kderusso reviewed (commented):
Changes look good, thanks!

Mikep86 committed:
SynonymsSource enum refactoring

Mikep86 committed:
Get SynonymsSource from settings via static method

Mikep86 committed:
Set lenient to true by default

mayya-sharipova commented on modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/SynonymTokenFilterFactory.java:
```diff
@@ -30,9 +30,57 @@
 
 public class SynonymTokenFilterFactory extends AbstractTokenFilterFactory {
+    protected enum SynonymsSource {
+        INLINE("synonyms", SynonymTokenFilterFactory::getRulesReader_Inline),
```
Should we have the standard camel case function names:

`getRulesReader_Inline` -> `getRulesReaderInline`?

Mikep86 committed:
Fix spotless violations

mayya-sharipova commented on modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/SynonymTokenFilterFactory.java:
```diff
@@ -62,8 +111,10 @@ public class SynonymTokenFilterFactory extends AbstractTokenFilterFactory {
+        this.synonymsSource = SynonymsSource.fromSettings(settings);
         this.expand = settings.getAsBoolean("expand", true);
-        this.lenient = settings.getAsBoolean("lenient", false);
+        this.lenient = settings.getAsBoolean("lenient", true);
```
Is this set `true` for all types of synonyms? From the title of PR it looks like it should `true` only for `synonyms_set` synonyms?

Mikep86 replied:
Yes, we decided to make that change after discussion in this PR and I haven't updated the PR name yet

mayya-sharipova reviewed (approved):
@Mikep86 Thanks, I left small comments, but overall LGTM.  +1 to merge after we are ok with breaking changes.

Mikep86 committed:
Merge branch 'main' into handle-invalid-synonyms-sets

Mikep86 committed:
Set default lenient value based on updateable value

Mikep86 committed:
Make lenient == true conditional on index version

Mikep86 committed:
Fix synonyms analysis tests

Mikep86 committed:
Converted synonyms API integration tests to YAML tests

Mikep86 committed:
Added invalid synonyms update test to ReloadSynonymAnalyzerIT

Mikep86 committed:
Fix spotless violations

Mikep86 committed:
Delete SynonymsApiIT

Mikep86 committed:
Rename reader methods

Mikep86 committed:
Merge branch 'main' into handle-invalid-synonyms-sets

Mikep86 committed:
Update changelog summary

Mikep86 commented:
@carlosdelest @kderusso @mayya-sharipova 

This PR is ready for re-review. Since last review, I've made the following changes:

- For 8.16.x indices, set `lenient` based on `updateable`
- Converted Synonyms API integration tests to YAML tests
- Updated synonyms analysis tests
- Added an invalid synonyms update test to `ReloadSynonymAnalyzerIT`
- Updated changelog

Documentation updates are still required. I'm seeking guidance on how to document breaking changes first.

kderusso reviewed (approved):
Updates look good to me, thanks for iterating! 

One question: Do we have to add a `>breaking` label to the PR too?

mayya-sharipova commented on server/src/main/java/org/elasticsearch/index/IndexVersions.java:
```diff
@@ -112,6 +112,7 @@ private static IndexVersion def(int id, Version luceneVersion) {
     public static final IndexVersion ADD_SECURITY_MIGRATION = def(8_510_00_0, Version.LUCENE_9_11_0);
     public static final IndexVersion UPGRADE_TO_LUCENE_9_11_1 = def(8_511_00_0, Version.LUCENE_9_11_1);
     public static final IndexVersion INDEX_SORTING_ON_NESTED = def(8_512_00_0, Version.LUCENE_9_11_1);
+    public static final IndexVersion LENIENT_UPDATEABLE_SYNONYMS = def(8_513_00_0, Version.LUCENE_9_11_1);
```
should this be adjusted for 8.15?

Mikep86 replied:
How would we adjust this for 8.16?

Mikep86 committed:
Update docs/changelog/110901.yaml

elasticsearchmachine commented:
Hi @Mikep86, I've updated the changelog YAML for you. *Note* that since this PR is labelled `>breaking`, you need to update the changelog YAML to fill out the extended information sections.

Mikep86 commented on docs/changelog/110901.yaml:
```diff
@@ -0,0 +1,12 @@
+area: Analysis
+type: "breaking, bug"
+issues: []
+breaking:
+  title: Gracefully handle invalid synonym rules in updateable synonyms
+  area: "Analysis, Relevance"
+  details: Please describe the details of this change for the release notes. You can
+    use asciidoc.
+  impact: Please describe the impact of this change to users
+  notable: false
```
Will update this tomorrow

Commit:
Include readers in enum instances

carlosdelest commented on modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/SynonymTokenFilterFactory.java:
```diff
@@ -152,46 +207,56 @@ SynonymMap buildSynonyms(Analyzer analyzer, ReaderWithOrigin rules) {
+        }
+    }
+
+    private static ReaderWithOrigin getRulesReaderInline(SynonymTokenFilterFactory factory, IndexCreationContext context) {
```
These methods could be inlined directly in the enum instances to make it explicit there's a change in behaviour in the different instances, see https://github.com/Mikep86/elasticsearch/pull/2. That way we get rid of the provider in the constructor which is clearer IMO

carlosdelest commented on modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/SynonymTokenFilterFactory.java:
```diff
@@ -30,9 +31,57 @@
 
 public class SynonymTokenFilterFactory extends AbstractTokenFilterFactory {
+    protected enum SynonymsSource {
+        INLINE("synonyms", SynonymTokenFilterFactory::getRulesReaderInline),
```
Enum instances are much clearer now, thanks!

carlosdelest reviewed (approved):
Great work @Mikep86 !

I'd like to have testing for the non-lenient (non default) behaviour, but the rest LGTM! 💯

Mikep86 commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/synonyms/80_synonyms_from_index.yml:
```diff
@@ -187,3 +187,44 @@ setup:
   - match: { shards_acknowledged: false }
+
+---
+"Load empty synonyms set from index for an analyzer":
```
No, but it wasn't explicitly tested and I wanted to add coverage since I removed the [fake rule](https://github.com/elastic/elasticsearch/blob/4173feefc2f23284cc2f151d7ff9807f2985f313/modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/SynonymTokenFilterFactory.java#L174) that was added previously.

carlosdelest replied:
Makes sense, thanks!

Mikep86 committed:
Fixed changelog

Mikep86 committed:
Added tests with lenient set to false

Mikep86 committed:
Merge pull request #2 from carlosdelest/handle-invalid-synonyms-sets-enum-changes

Handle synonym sets: Include readers in enum instances

Mikep86 committed:
Updated synonym & synonym graph token filter documentation

elasticsearchmachine commented:
Pinging @elastic/es-docs (Team:Docs)

Mikep86 committed:
Delete docs/changelog/110901.yaml

Mikep86 committed:
Update docs/changelog/110901.yaml

elasticsearchmachine commented:
Hi @Mikep86, I've created a changelog YAML for you. *Note* that since this PR is labelled `>breaking`, you need to update the changelog YAML to fill out the extended information sections.

Mikep86 committed:
Update changelog

Mikep86 commented on docs/reference/analysis/tokenfilters/synonym-graph-tokenfilter.asciidoc:
@kderusso @carlosdelest @mayya-sharipova Can I get some 👀 on these doc changes?

carlosdelest replied:
I have a comment on the note that includes the release version, the rest LGTM 👍

carlosdelest commented on docs/reference/analysis/tokenfilters/synonym-graph-tokenfilter.asciidoc:
```diff
@@ -87,11 +87,17 @@ changes to synonym files. Only to be used for search analyzers.
 It is important to note that only those synonym rules which cannot get parsed are ignored.
 See <<synonym-graph-tokenizer-stop-token-filter,synonyms and stop token filters>> for an example of `lenient` behaviour for invalid synonym rules.
 
+[INFO]
```
I like the extra information, but I'm thinking that this should be part of release notes and it will quickly get old (i.e. this will still be around in 9.x).

Maybe docs team has more guidance on this. I wouldn't include it as I don't see this kind of advice in other parts of our docs.

benwtrent replied:
I agree, this is what release notes and breaking changes are for. This callout should be removed.

kderusso replied:
++, let's remove the version dependent callout.

If you feel that we should have more information in our documentation, you could add information on the `lenient` parameter that if lenient is set to true invalid synonyms will be ignored/skipped. That's potentially a little more clear than "errors are ignored" That might get at the underlying concern but still provide non-stale information to users.

It could be good to backport the warnings about lenient set to true to older branches.

Mikep86 replied:
Ok, I will remove this callout. This information is already in the release notes per the changelog.

@kderusso Carlos already added a lot of info about the `lenient` parameter in #110822, which was backported all the way back to 8.10. So I think we're good there.

kderusso reviewed (approved):
Overall LGTM, agreed on the version specific verbiage and one minor suggestion.

Mikep86 committed:
Remove callout
